### PR TITLE
memtx: fix self-conflicting of transactions

### LINKED ISSUE #7217: Repeatable {in, up}sert with memtx transaction manager enabled causes transaction conflict
**Steps to reproduce**
```lua
#!/usr/bin/env tarantool

os.execute('rm -rf *.snap *.xlog *.vylog 512')

local log = require('log')
local txn_proxy = require('txn_proxy')
local yaml = require('yaml')

box.cfg{memtx_use_mvcc_engine = true}

box.schema.space.create('s')
box.space.s:create_index('pk')

local tx1 = txn_proxy:new()
tx1('box.begin()')

local tx2 = txn_proxy:new()
tx2('box.begin()')

tx1('box.space.s:replace{0, 0}')

tx2('box.space.s:replace{0, 1}')
tx2('box.space.s:delete{0}')
tx2('box.space.s:insert{0, 2}')

tx1('box.commit()')
log.info("%s", yaml.encode(tx2('box.commit()')))

os.exit()
```

```lua
#!/usr/bin/env tarantool

os.execute('rm -rf *.snap *.xlog *.vylog 512')

local log = require('log')
local txn_proxy = require('txn_proxy')
local yaml = require('yaml')

box.cfg{memtx_use_mvcc_engine = true}

box.schema.space.create('s')
box.space.s:create_index('pk')

local tx1 = txn_proxy:new()
tx1('box.begin()')

local tx2 = txn_proxy:new()
tx2('box.begin()')

tx1('box.space.s:replace{0, 0}')

tx2('box.space.s:replace{0, 1}')
tx2('box.space.s:delete{0}')
tx2('box.space.s:upsert({0, 2}, {{"=", 2, 2}})')

tx1('box.commit()')
log.info("%s", yaml.encode(tx2('box.commit()')))

os.exit()
```

**Actual behavior**
```console
 ---
- {'error': 'Transaction has been aborted by conflict'}
...
```

**Expected behavior**
```console
 ---
...
```

### LINKED ISSUE #7221: {Replace, {in, up}sert} after full scan in hash index with memtx transaction manager enabled causes transaction to self-conflict
**Bug description**
Consider the following fair assertion:
```diff
diff --git a/src/box/memtx_tx.c b/src/box/memtx_tx.c
index 74814cfbe..d32128e6d 100644
--- a/src/box/memtx_tx.c
+++ b/src/box/memtx_tx.c
@@ -543,6 +543,7 @@ memtx_tx_abort_all_for_ddl(struct txn *ddl_owner)
 int
 memtx_tx_cause_conflict(struct txn *breaker, struct txn *victim)
 {
+       assert(breaker != victim);
        struct tx_conflict_tracker *tracker = NULL;
        struct rlist *r1 = breaker->conflict_list.next;
        struct rlist *r2 = victim->conflicted_by_list.next;
```
**Steps to reproduce**
```lua
os.execute('rm -rf *.snap *.xlog *.vylog 512')

local txn_proxy = require('txn_proxy')
local log = require('log')
local yaml = require('yaml')

box.cfg{memtx_use_mvcc_engine = true}

box.schema.space.create('s')
box.space.s:create_index('pk', {type = 'hash'})

local tx1 = txn_proxy.new()
tx1:begin()
tx1('box.space.s:select{}')
tx1('box.space.s:replace{0}')
log.info("%s", yaml.encode(tx1:commit()))

os.exit()
```

```lua
os.execute('rm -rf *.snap *.xlog *.vylog 512')

local txn_proxy = require('txn_proxy')
local log = require('log')
local yaml = require('yaml')

box.cfg{memtx_use_mvcc_engine = true}

box.schema.space.create('s')
box.space.s:create_index('pk', {type = 'hash'})

local tx1 = txn_proxy.new()
tx1:begin()
tx1('box.space.s:select{}')
tx1('box.space.s:insert{0}')
log.info("%s", yaml.encode(tx1:commit()))

os.exit()
```

```lua
os.execute('rm -rf *.snap *.xlog *.vylog 512')

local txn_proxy = require('txn_proxy')
local log = require('log')
local yaml = require('yaml')

box.cfg{memtx_use_mvcc_engine = true}

box.schema.space.create('s')
box.space.s:create_index('pk', {type = 'hash'})

local tx1 = txn_proxy.new()
tx1:begin()
tx1('box.space.s:select{}')
tx1('box.space.s:upsert({0}, {{"=", 1, 0}})')
log.info("%s", yaml.encode(tx1:commit()))

os.exit()
```
**Actual behavior**
```console
Assertion failed: (breaker != victim), function memtx_tx_cause_conflict, file memtx_tx.c, line 546.
```
**Expected behavior**
Script successfully exits.

### LINKED ISSUE #7234: Space truncate with memtx transaction manager enabled causes transaction to self-conflict
**Bug description**
Consider the following fair assertion:
```diff
diff --git a/src/box/memtx_tx.c b/src/box/memtx_tx.c
index 74814cfbe..d32128e6d 100644
--- a/src/box/memtx_tx.c
+++ b/src/box/memtx_tx.c
@@ -543,6 +543,7 @@ memtx_tx_abort_all_for_ddl(struct txn *ddl_owner)
 int
 memtx_tx_cause_conflict(struct txn *breaker, struct txn *victim)
 {
+       assert(breaker != victim);
        struct tx_conflict_tracker *tracker = NULL;
        struct rlist *r1 = breaker->conflict_list.next;
        struct rlist *r2 = victim->conflicted_by_list.next;
```
**Steps to reproduce**
```lua
os.execute('rm -rf *.snap *.xlog *.vylog 512')

box.cfg{memtx_use_mvcc_engine = true}

box.schema.space.create('s')
box.schema.space.s:truncate()

os.exit()
```
**Actual behavior**
```console
Assertion failed: (breaker != victim), function memtx_tx_cause_conflict, file memtx_tx.c, line 546.
```
**Expected behavior**
Script successfully exits.

### LINKED ISSUE #7235: {Replace, {in, up}sert} after selecting the same key in TREE index with memtx transaction manager enabled causes transaction to self-conflict
**Bug description**
Consider the following fair assertion:
```diff
diff --git a/src/box/memtx_tx.c b/src/box/memtx_tx.c
index 74814cfbe..d32128e6d 100644
--- a/src/box/memtx_tx.c
+++ b/src/box/memtx_tx.c
@@ -543,6 +543,7 @@ memtx_tx_abort_all_for_ddl(struct txn *ddl_owner)
 int
 memtx_tx_cause_conflict(struct txn *breaker, struct txn *victim)
 {
+       assert(breaker != victim);
        struct tx_conflict_tracker *tracker = NULL;
        struct rlist *r1 = breaker->conflict_list.next;
        struct rlist *r2 = victim->conflicted_by_list.next;
```
**Steps to reproduce**
```lua
os.execute('rm -rf *.snap *.xlog *.vylog 512')

local txn_proxy = require('txn_proxy')

box.cfg{memtx_use_mvcc_engine = true}

box.schema.space.create('s')
box.space.s:create_index('pk')

local tx = txn_proxy:new()

tx:begin()
tx('box.space.s:select{0}')
tx('box.space.s:replace{0}')
tx:commit()

os.exit()
```

```lua
os.execute('rm -rf *.snap *.xlog *.vylog 512')

local txn_proxy = require('txn_proxy')

box.cfg{memtx_use_mvcc_engine = true}

box.schema.space.create('s')
box.space.s:create_index('pk')

local tx = txn_proxy:new()

tx:begin()
tx('box.space.s:select{0}')
tx('box.space.s:insert{0}')
tx:commit()

os.exit()
```

```lua
os.execute('rm -rf *.snap *.xlog *.vylog 512')

local txn_proxy = require('txn_proxy')

box.cfg{memtx_use_mvcc_engine = true}

box.schema.space.create('s')
box.space.s:create_index('pk')

local tx = txn_proxy:new()

tx:begin()
tx('box.space.s:select{0}')
tx('box.space.s:upsert({0}, {{"=", 1, 0}})')
tx:commit()

os.exit()
```
**Actual behavior**
```console
Assertion failed: (breaker != victim), function memtx_tx_cause_conflict, file memtx_tx.c, line 546.
```
**Expected behavior**
Script successfully exits.

### PR DESCRIPTION
**Pull request by CuriousGeorgiy**
### memtx: fix DML causing transaction self-conflict after full scan in HASH

When full scans are checked on writes, we must only conflict transactions
other than the one that did the full scan.

Closes #7221

### memtx: fix DML causing transaction self-conflict after selecting same key

On insertion, when point holes are checked on insertion, we must only
conflict transactions other than the one that read the hole.

NO_CHANGELOG=internal bugfix
NO_DOC=bugfix

Closes #7234
Closes #7235

### memtx: add assertions for self-conflicting of transactions

We assume that a transaction cannot conflict itself and that only a
non-prepared transaction can be conflicted: track these assumptions with
appropriate assertions.

### CODE REVIEW

**Review by drewdzzz — CHANGES_REQUESTED** [review]
It's all good except for one thing: it's better to write a changelog, at least one for the entire patch, because you changed the behavior of the transaction manager.

**Review by drewdzzz — APPROVED** [review]

File: src/box/memtx_tx.c [hunk 1]
```diff
@@ -1499,7 +1499,8 @@ check_hole(struct space *space, uint32_t index,
```
Comment (alyapunov) [inline src/box/memtx_tx.c:1503]: Could you please make the title of the commit a bit shorter? It is shown very strange in Web (an I bet somewhere else).

**Review by alyapunov — APPROVED** [review]

File: src/box/memtx_tx.c (hunk 1 above)
Comment (CuriousGeorgiy) [reply src/box/memtx_tx.c:1503]: Fixed: shortened the commit subject.

@locker seems like a bug in checkpatch (tarantool/checkpatch#18).

**Comment by locker** [comment]
Cherry-picked to 2.10.

**Pull request merged**